Assignment4.java

```diff
@@ -0,0 +1,117 @@
+                    numberOfIslands++;
+                }
+            }
+        }
+
+        return numberOfIslands;
+    }
+
+    /**
+     * Discovers an island when given a part of it.
+     *
+     * @param row     row number of current tile in map
+     * @param column  column number of current tile n map
+     * @param map     map of the area
+     * @param visited holds whether a tile is visited or not
+     */
+    private static void discoverIsland(int row, int column, boolean[][] map, boolean[][] visited) {
+        visited[row][column] = true;
+
+        // check up tile
+        if (row > 0 && map[row - 1][column] && !visited[row - 1][column]) {
+            discoverIsland(row - 1, column, map, visited);
+        }
+
+        // check down tile
+        if (row < map.length - 1 && map[row + 1][column] && !visited[row + 1][column]) {
+            discoverIsland(row + 1, column, map, visited);
+        }
+
+        // check left tile
+        if (column > 0 && map[row][column - 1] && !visited[row][column - 1]) {
+            discoverIsland(row, column - 1, map, visited);
+        }
+
+        // check right tile
+        if (column < map[0].length - 1 && map[row][column + 1] && !visited[row][column + 1]) {
+            discoverIsland(row, column + 1, map, visited);
+        }
+    }
+
+    /**
+     * Finds the number of islands in the map using disjoint set.
+     *
+     * @param rowNumber    number of rows
+     * @param columnNumber number of columns
+     * @param map          map of the area
+     * @return number of islands, 0 if map is null
+     */
+    public static int findNumberOfIslandsDisjointSet(int rowNumber, int columnNumber, boolean[][] map) {
+        if (map == null) {
+            return 0;
+        }
+
+        DisjointSet disjointSet = new DisjointSet(rowNumber * columnNumber);
+        boolean isLand;
+
+        for (int i = 0; i < rowNumber; i++) {
+            for (int j = 0; j < columnNumber; j++) {
+                isLand = map[i][j];
+
+                if (isLand) {
+                    disjointSet.makeSet(findId(i, j, columnNumber));
+
+                    // union with land above tile
+                    if (i > 0 && map[i - 1][j]) {
+                        disjointSet.union(findId(i, j, columnNumber), findId(i - 1, j, columnNumber));
+                    }
+
+                    // union with land to the left of tile
+                    if (j > 0 && map[i][j - 1]) {
+                        disjointSet.union(findId(i, j, columnNumber), findId(i, j - 1, columnNumber));
+                    }
+                }
+            }
+        }
+
+        return disjointSet.getCount();
+
+    }
+
+    /**
+     * Calculates the id of a tile in the map.
+     * Ids of tiles start from 1 and go up by 1 accordingly.
+     *
+     * @param rowNumber       row number of the tile
+     * @param columnNumber    column number of the tile
+     * @param mapColumnNumber column number of the map
```

Comment:
You might consider renaming variables to make distinction between index and count more clear, for example use just `row` and `column` for indexes, reserving names with Number to represent "number of columns/rows". Other options might include `numberOfColumns` as clearly different from `columnNumber` etc.

If you rename them, it would be good to keep names in other methods changed too, so that naming is consistent.